File: broken-object-level-authorization/go/secure_version.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"time"
)

type User struct {
	ID       int    `json:"id"`
	Username string `json:"username"`
	Email    string `json:"email"`
}

var users = map[int]User{
	1: {ID: 1, Username: "user1", Email: "user1@example.com"},
	2: {ID: 2, Username: "user2", Email: "user2@example.com"},
}

func getUser(w http.ResponseWriter, r *http.Request) {
	// Read Header x-user-id simulating user authentication
	authenticatedUserIDStr := r.Header.Get("x-user-id")
	authenticatedUserID, err := strconv.Atoi(authenticatedUserIDStr)
	if err != nil {
		http.Error(w, "Invalid user ID", http.StatusBadRequest)
		return
	}

	userIDStr := r.URL.Query().Get("id")
	userID, err := strconv.Atoi(userIDStr)
	if err != nil {
		http.Error(w, "Invalid user ID", http.StatusBadRequest)
		return
	}

	// ------------ Verificación de autorización ------------
	if userID != authenticatedUserID {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	user, exists := users[userID]
	if !exists {
		http.Error(w, "User not found", http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(user); err != nil {
        http.Error(w, "Failed to encode user data", http.StatusInternalServerError)
    }
}

func main() {
	http.HandleFunc("/user", getUser)
	fmt.Println("Go Secure Version Server Broken Object Level Authorization started at :8080")
	server := &http.Server{
        Addr:         ":8080",
        ReadTimeout:  5 * time.Second,
        WriteTimeout: 10 * time.Second,
        IdleTimeout:  15 * time.Second,
    }
	if err := server.ListenAndServe(); err != nil {
        fmt.Printf("Server failed to start")
    }
}

File: broken-object-level-authorization/node/secure_version.js
const express = require('express');
const app = express();
const port = 8080;

// Datos de usuarios
const users = {
    1: { id: 1, username: "user1", email: "user1@example.com" },
    2: { id: 2, username: "user2", email: "user2@example.com" }
};

// Middleware para parsear el header x-user-id
app.use((req, res, next) => {
    const authenticatedUserID = req.headers['x-user-id'];
    if (!authenticatedUserID) {
        return res.status(400).json({ error: "Invalid user ID" });
    }
    req.authenticatedUserID = parseInt(authenticatedUserID, 10);
    if (isNaN(req.authenticatedUserID)) {
        return res.status(400).json({ error: "Invalid user ID" });
    }
    next();
});

// Ruta para obtener un usuario
app.get('/user', (req, res) => {
    const authenticatedUserID = req.authenticatedUserID;

    const userID = parseInt(req.query.id, 10);
    if (isNaN(userID)) {
        return res.status(400).json({ error: "Invalid user ID" });
    }

    // ------------ Verificación de autorización ------------
    if (userID !== authenticatedUserID) {
        return res.status(401).json({ error: "Unauthorized" });
    }

    const user = users[userID];
    if (!user) {
        return res.status(404).json({ error: "User not found" });
    }

    res.json(user);
});

// Configuración del servidor con timeouts
const server = app.listen(port, () => {
    console.log(`Node Secure Version Server Broken Object Level Authorization started at http://localhost:${port}`);
});

// Configuración de timeouts
server.keepAliveTimeout = 15000; // 15 segundos
server.headersTimeout = 10000; // 10 segundos